feat(zkevm-circuits): enable type 1 and 2 as an input data

dongchangYoo opened the pull request:
There are 2 changes as below:

feat(zkevm-circuits): enable type 1 and 2 as a input data
- fix bug on handling access list
- ensure that the address in access list and tx's 'from/to'
  are only written once to state db in circuit input builder.
  If not, `Lookup Rw` error occurs

refac(integration-test): refactoring integration test
- The content of the gendata_output file is guaranteed.
  (If txs are not in a block, it will be dead.)
- reduced the amount of reusable code.
- simplify the setup process logs.
- add test case for type 2 tx with access list

dongchangYoo commented:
Due to a minor oversight, a few forced pushes were performed. As there are no further tasks remaining, I kindly request a review.

fakedev9999 commented:
@dongchangYoo for the third commit message, let's fix "refactoring" to "refactor", following our convention to use imperative mood.

fakedev9999 commented on zkevm-circuits/src/super_circuit.rs:
```diff
@@ -972,7 +1018,7 @@ pub(crate) mod super_circuit_tests {
     fn serial_test_super_circuit_2tx_2max_tx() {
-        let block = block_2tx();
+        let block = block_type2_1tx();
         const MAX_TXS: usize = tx_idx!(2);
```
```suggestion
        const MAX_TXS: usize = tx_idx!(1);
```

fakedev9999 commented on integration-tests/src/bin/gen_blockchain_data.rs:
```diff
@@ -156,240 +49,162 @@ async fn main() {
-        "Deploy Greeter".to_string(),
-        receipt.block_number.unwrap().as_u64(),
-    );
+    // Deploy "Greter" contract
```
```suggestion
    // Deploy "Greeter" contract
```

fakedev9999 commented on integration-tests/src/bin/gen_blockchain_data.rs:
```diff
@@ -156,240 +49,162 @@ async fn main() {
-    );
+    // Deploy "Greter" contract
+    let contract_name = "Greeter";
+    let target_contarct = contracts.get(contract_name).expect("contarct not found");
```
```suggestion
    let target_contract = contracts.get(contract_name).expect("contract not found");
```

fakedev9999 commented on integration-tests/src/bin/gen_blockchain_data.rs:
```diff
@@ -156,240 +49,162 @@ async fn main() {
-        "Deploy OpenZeppelinERC20TestToken".to_string(),
-        receipt.block_number.unwrap().as_u64(),
-    );
+    // Deploy "ERC20" contract
```
```suggestion
    // Deploy "OpenZeppelinERC20TestToken" contract
```

TomTaehoonKim commented on integration-tests/src/bin/gen_blockchain_data.rs:
```diff
@@ -156,240 +49,162 @@ async fn main() {
-        receipt.block_number.unwrap().as_u64(),
-    );
+    // Deploy "Greter" contract
+    let contract_name = "Greeter";
```
```suggestion
    let contract_name = "Greeter";
```

fakedev9999 commented on integration-tests/src/bin/gen_blockchain_data.rs:
```diff
@@ -156,240 +49,162 @@ async fn main() {
-    );
+    // Deploy "ERC20" contract
+    let contract_name = "OpenZeppelinERC20TestToken";
+    let target_contarct = contracts.get(contract_name).expect("contarct not found");
```
```suggestion
    let target_contract = contracts.get(contract_name).expect("contract not found");
```

fakedev9999 commented on integration-tests/src/bin/gen_blockchain_data.rs:
```diff
@@ -156,240 +49,162 @@ async fn main() {
         "Transfer 0".to_string(),
         receipt.block_number.unwrap().as_u64(),
     );
+    info!("- done(height: {:?})", receipt.block_number);
```
```suggestion
    info!("- Done(height: {:?})", receipt.block_number);
```
Can you apply all the changes to Done? I suggest this change because we use "Doing" capitalized.

fakedev9999 commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+}
+
+/// Build and return erc20 tranfer transaction
+/// TODO(dongchangYoo) Currently this function does not support that an account sends tx twice in a
```
```suggestion
/// TODO(dongchangYoo): Currently this function does not support sending multiple txs from an account in a
```

fakedev9999 commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+
+use crate::{get_provider, get_wallet, CompiledContract, CONTRACTS, CONTRACTS_PATH};
+
+/// Number of transaction in each scenario
```
```suggestion
/// Number of transactions in each scenario
```

fakedev9999 commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+    prov
+}
+
+/// init wallets to be used in each scenario
```
```suggestion
/// Init wallets to be used in each scenario
```

fakedev9999 commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+    contracts
+}
+
+/// Depoloy contract
```
```suggestion
/// Deploy contract
```

fakedev9999 commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+    (receipt.block_number.unwrap().as_u64(), contract)
+}
+
+/// An `sender` account sends ETH to multiple accounts
```
```suggestion
/// A `sender` account sends ETH to multiple accounts
```

fakedev9999 commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+        let call = call.gas(100_000);
+
+        // Convert tx to legacy tx if tx_type equals to 0
+        // TODO(dongchangYoo) use constant for tx_type after implementing various tx types
```
```suggestion
        // TODO(dongchangYoo): use constant for tx_type after implementing various tx types
```

fakedev9999 commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+/// Build and return erc20 tranfer transaction
+/// TODO(dongchangYoo) Currently this function does not support that an account sends tx twice in a
+/// function.
+pub async fn exchange_token(
```
```suggestion
pub async fn transfer_erc20_token(
```
Or is the "exchange" a formal term to use in a situation of ERC-20 token transfers?

chokobole commented on .gitignore:
```diff
@@ -5,4 +5,6 @@
-*.sh
\ No newline at end of file
+*.sh
+*.h
+*.so
```
Why did you add this?

chokobole commented on zkevm-circuits/src/super_circuit.rs:
```diff
@@ -869,6 +869,52 @@ pub(crate) mod super_circuit_tests {
         block
     }
 
+    pub(crate) fn block_type2_1tx() -> GethData {
```
For me, it's better to name `dynamic fee tx` rather than `type2 tx`.

```suggestion
    pub(crate) fn block_1dynamic_fee_tx() -> GethData {
```

chokobole commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+        )
+    }
+
+    sleep(Duration::from_secs(5));
```
Please remove this!

chokobole commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+
+        // Convert tx to legacy tx if tx_type equals to 0
+        // TODO(dongchangYoo) use constant for tx_type after implementing various tx types
+        let tx = match tx_type {
+            0 => {
+                let call = call.legacy();
+                call.tx
+            }
+            _ => call.tx,
+        };
```
```suggestion
        let tx = if tx_type == 0 { 
            let call = call.legacy();
            call.tx
        } else {
            call.tx
        };
```

chokobole commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+            wallets[trans_ctx.from_i]
+                .send_transaction(tx, None)
+                .await
+                .expect("cannot send token"),
```
```suggestion
                .expect("cannot transfer token"),
```

chokobole replied:
Above, you used "cannot send tx" for the error message. Please unify the error message.

chokobole commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+
+/// TransferContext
+pub struct TransferContext {
+    /// Index of `from` wallet
+    pub from_i: usize,
+    /// Index of `to` wallet
+    pub to_i: usize,
+    /// Amount to transfer
+    pub amount: U256,
+    /// Expected result
+    pub expected_result: bool,
```
Let's simplify the names

```suggestion
    /// Index of `from` wallet
    pub from: usize,
    /// Index of `to` wallet
    pub to: usize,
    /// Amount to transfer
    pub amount: U256,
    /// Expected result
    pub expected: bool,
```

chokobole commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+        nonce = nonce.checked_add(U256::one()).unwrap();
+    }
+
+    sleep(Duration::from_secs(5));
```
Please remove this!

chokobole commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+    }
+
+    sleep(Duration::from_secs(5));
+    block_num = 0;
+    for tx in pending_txs.into_iter() {
+        let receipt = tx.await.expect("cannot confirm tx").unwrap();
+        if block_num == 0 {
+            block_num = receipt.block_number.unwrap().as_u64();
+        } else if block_num != receipt.block_number.unwrap().as_u64() {
+            panic!("The txs are not in a block")
+        }
+    }
+    block_num
```
Please extract this to a function and do not repeat this.

chokobole commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,236 @@
+    let factory = ContractFactory::new(compiled.abi.clone(), compiled.bin.clone(), wallet.clone());
+    let (contract, receipt) = factory
+        .deploy(args)
+        .unwrap_or_else(|_| panic!("cannot deploy: {:?}", compiled.name))
```
```suggestion
        .unwrap_or_else(|_| panic!("cannot construct contract creation tx: {:?}", compiled.name))
```

chokobole commented:
How about changing commit titles like below?

- `build(.gitignore): change gitignore` -> `chore: change gitignore`
- `feat(zkevm-circuits): enable type 1 and 2 as a input data` -> `feat(zkevm-circuits): enable type 1 and 2 as an input data`
- `refac(integration-test): refactoring integration test` -> `test(integration-test): improve integration test`

dongchangYoo committed:
fix(zkevm-circuits): fix bug on handling access list

ensure that the address in access list and tx's 'from/to'
are only written once to state db in circuit input builder.
If not, `Lookup Rw` error occurs.

dongchangYoo committed:
feat(zkevm-circuits): support type 1,2 tx (1/2)

In order to validate type 1 and type 2 transactions,
the following two must be implemented:
- Verification of type-specific RLP encoding.
- Application of the access list concept.
This commit addresses the second item, while the first item
will be handled in the subsequent process.

dongchangYoo committed:
refac(integration-test): refactor integration test

chokobole commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,223 @@
+        .confirmations(0usize)
+        .send_with_receipt()
+        .await
+        .unwrap_or_else(|_| panic!("cannot construct contract creation tx: {:?}", compiled.name));
```
```suggestion
        .unwrap_or_else(|_| panic!("cannot deploy tx: {:?}", compiled.name));
```

fakedev9999 commented on integration-tests/src/bin/gen_blockchain_data.rs:
```diff
@@ -156,240 +49,162 @@ async fn main() {
+    )
+    .await;
+    blocks.insert(case_name.to_string(), block_num);
+    info!("- done(height: {:?})", block_num);
```
```suggestion
    info!("- Done(height: {:?})", block_num);
```

fakedev9999 commented on integration-tests/src/scenario_utils.rs:
```diff
@@ -0,0 +1,223 @@
+
+/// Build and return erc20 transfer transaction
+/// TODO(dongchangYoo): Currently this function does not support sending multiple txs from an
+/// account in a
```
```suggestion
/// account.
```
Sorry, I gave you an incomplete suggestion before.

lightscale-luke reviewed (approved):
LGTM

chokobole reviewed (approved):
LGTM

fakedev9999 reviewed (approved):
Good job!! Looks good to go~
